op-node/sources/sync_client.go

```diff
@@ -0,0 +1,116 @@
+package sources
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/ethereum-optimism/optimism/op-node/client"
+	"github.com/ethereum-optimism/optimism/op-node/eth"
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-node/sources/caching"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type SyncClientInterface interface {
+	Start(unsafeL2Payloads chan *eth.ExecutionPayload) error
+	Close() error
+	fetchUnsafeBlockFromRpc(ctx context.Context, blockNumber uint64)
+}
+
+type SyncClient struct {
+	*L2Client
+	FetchUnsafeBlock chan uint64
+	done             chan struct{}
+	unsafeL2Payloads chan *eth.ExecutionPayload
+	wg               sync.WaitGroup
+}
+
+var _ SyncClientInterface = (*SyncClient)(nil)
+
+type SyncClientConfig struct {
+	L2ClientConfig
+}
+
+func SyncClientDefaultConfig(config *rollup.Config, trustRPC bool) *SyncClientConfig {
+	return &SyncClientConfig{
+		*L2ClientDefaultConfig(config, trustRPC),
+	}
+}
+
+func NewSyncClient(client client.RPC, log log.Logger, metrics caching.Metrics, config *SyncClientConfig) (*SyncClient, error) {
+	l2Client, err := NewL2Client(client, log, metrics, &config.L2ClientConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &SyncClient{
+		L2Client:         l2Client,
+		FetchUnsafeBlock: make(chan uint64),
```

Comment:
This is an unbuffered channel so i think it'll block the driver event loop as its doing work over here, it would be nice to have metrics on how long it takes to fetch each block

Reply:
Buffered the channel at `128` and enforced a max gap between the unsafe head / payload queue tip to be `128` as well. Does this number seem reasonable @protolambda @trianglesphere? A max gap size of `128` accounts for roughly 4 1/4 minutes worth of L2 blocks.

Reply:
Also, should this max gap size (as well as the buffered channel size) be configurable for people that are syncing from a backup RPC that they own and/or know won't rate limit them?

Reply:
why does the gap & channel buffer length need to be the same? Are you actually storing blocks inside the channel?

Reply:
@trianglesphere The channel stores `uint64`s, which represent block numbers- Correct me if I'm wrong, but for a max gap of 128, the `FetchUnsafeBlock` channel will need to store 128 block numbers at max?